[issue-839] Drop Python 3.7 support and fix CI pipeline issues

armintaenzertng opened the pull request:
fixes #839 
after some updates, this also fixes #844 

ubuntu-latest does not support Python 3.7 anymore.
See https://github.com/actions/runner-images/issues/10636

armintaenzertng commented:
@kestewart, it's not about upgrading our Python version (we already support at least 3.11) but about whether we want to drop Python 3.7 support.
Python 3.7 reached end-of-life on 2023-06-27, but still seems to be in use by some people.

Note that we wouldn't immediately lose compatibility when erasing it from the pipeline, but we wouldn't catch changes in our code that break for Python 3.7, so we couldn't guarantee 3.7 support anymore.

maxhbr commented:
I think we should drop support for python versions that are EOL

kestewart commented:
@armintaenzertng, @maxhbr - no issue with dropping support - what I was objecting to was the pegging to an old version of ubuntu, rather than staying with the latest.     If I've misunderstood what the patch is doing, please explain.

maxhbr commented:
good point, lets drop old versions and stay with latest ubuntu

armintaenzertng commented:
Sorry for the slow response. In order to get the pipeline running again I had to do several fixes:
- No Python 3.7 support anymore, as discussed
- Updated the exception raised by the type checking due to a new version of beartype
- Updated formatting due to a new version of black

@maxhbr, could you please remove the Python 3.7 checks from this repo? :)

meretp reviewed (approved):
lgtm, nit: you could update the PRs description

meretp commented:
btw 3.8 is also eol, isn't? we could drop that as well and should check if we support 3.12 and 3.13

meretp commented on src/spdx_tools/common/typing/dataclass_with_properties.py:
```diff
@@ -4,7 +4,7 @@
 from dataclasses import dataclass
 
 from beartype import beartype
-from beartype.roar import BeartypeCallHintException
+from beartype.roar import BeartypeCallHintParamViolation
 
```
does this relate to #844, does this change fix the issue mentioned there?

armintaenzertng commented:
I don't see an urgent need to drop support for 3.8 as long as it doesn't cost us anything. But I agree that we should check if we can add support for 3.12 and 3.13.
